### ISSUE TITLE
Fix npm shrinkwrap

### ISSUE THREAD

**Issue by garethbowen** [medic/medic-webapp#3032]
The way we're using npm shrinkwrap isn't working, or npm shrinkwrap is just broken.

* It's painful to add new dependencies because that regenerates the shrinkwrap which can break the build in unrelated ways
* It always tries to include the dev dependencies which are irrelevant and makes it difficult to diff
* It gets out of date partially because it's just too difficult to use properly

Things to consider...

1. Investigate bumping to npm v4 which is included in node v7.4 and has changed how shrinkwrap works so it might resolve the issues.
2. Investigate using [yarn](https://code.facebook.com/posts/1840075619545360), facebook's attempt to improve on npm.
3. Consider building the shrinkwrap as a build artefact so we can rebuild an exact published version if necessary but it doesn't get in our way at dev time.

**Comment by garethbowen** [comment]
Switching to yarn is looking more and more appealing.

**Comment by garethbowen** [comment]
NB: gardener which we use to do deployment does an `npm install` at install time. This probably makes this issue infeasible to do until we switch to Horti in 3.0.0 so I'll schedule it for then.

**Comment by rmhowe** [comment]
Ok so it looks like the current state of package managers is as follows:

**npm** (>=v5):
- Deterministic (same resulting `node_modules` from the same `package-lock.json`) across different npm versions
- Easy to resolve `package-lock.json` conflicts as of v5.7 using `npm install --package-lock-only`
- `nsp` support
- Can theoretically update deep dependencies using [`replace-deep-dep`](https://www.npmjs.com/package/replace-deep-dep) however this package doesn't seem to be very widely used/supported/maintained

**Yarn**:
- Deterministic (same resulting `node_modules` from the same `yarn.lock`) ONLY for the same yarn version, more info [here](https://yarnpkg.com/blog/2017/05/31/determinism/). Seems we would need to use the `engines` field in `package.json` to ensure everyone was on the same version of yarn
- Easy to resolve `yarn.lock` conflicts as of v1.0 using `yarn install`
- `nsp` support
- Can update deep dependencies using [`resolutions`](https://yarnpkg.com/en/docs/selective-version-resolutions)

As far as I can tell both package managers should satisfy most of our requirements (although there have been some problems with npm, these may have been to do with merging/updating `package-lock.json` incorrectly), except yarn's `resolutions` seem like they will work more reliably than npm's `replace-deep-dep`, but it's hard to say. Both include dev dependencies in their respective lock files. Also if we do use yarn, we'll need to restrict the version by using the `engines` field in `package.json`, but this seems acceptable to me so I'd vote for yarn.

cc @SCdF @alxndrsn @dianabarsan @vimemo

**Comment by garethbowen** [comment]
Yarn gets my vote too.

There are other issues with npm (random package integrity failures, found to be downloading versions other than those specified in the package-lock, etc) that have sapped my confidence in the product.

I'm also curious about performance as npm seems unnecessarily slow - can you do some profiling with our project on npm vs yarn?

**Comment by rmhowe** [comment]
I did a bunch of tests across webapp, api and sentinel using npm v5.8.0 and yarn v1.5.1

**Webapp**
| | npm | yarn |
| --- | --- | --- |
| Initial install without node_modules, lockfile, or cache | 73s | 69s |
| Initial install without node_modules or lockfile, with cache | 52s | 48s |
| Initial install without node_modules, with lockfile and cache | 33s | 32s |
| Installing `supports-colors` | 13s | 7s |
| Installing `react` | 14s | 9s |

**API**
| | npm | yarn |
| --- | --- | --- |
| Initial install without node_modules, lockfile, or cache | 17s | 16s |
| Initial install without node_modules or lockfile, with cache | 13s | 10s |
| Initial install without node_modules, with lockfile and cache | 9s | 5s |
| Installing `supports-colors` | 4s | 2s |
| Installing `react` | 5s | 5s |

**Sentinel**
| | npm | yarn |
| --- | --- | --- |
| Initial install without node_modules, lockfile, or cache | 10s | 10s |
| Initial install without node_modules or lockfile, with cache | 8s | 6s |
| Initial install without node_modules, with lockfile and cache | 5s | 3s |
| Installing `supports-colors` | 3s | 2s |
| Installing `react` | 4s | 4s |

**TL;DR** In webapp (with by far the biggest `package.json`) Yarn was pretty consistently faster but only by about 5-10%. In api and sentinel the times are so small there's not that much point in comparing, but again Yarn was slightly faster.

**Comment by ngaruko** [comment]
Does this need any AT @garethbowen @rmhowe apart from checking that the app can be deployed properly with `yarn`?

**Comment by garethbowen** [comment]
@ngaruko That'd be enough I think.

**Comment by ngaruko** [comment]
LGTM
